### PR TITLE
Cluster diagram for K-Means model summary

### PR DESCRIPTION
**Pull request by fazlan-nazeem**
https://wso2.org/jira/browse/ML-52

### CODE REVIEW

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java [hunk 1]
```diff
@@ -159,7 +168,38 @@ private ModelSummary buildKMeansModel(long modelID, JavaRDD<Vector> trainingData
 //            if (testingData != null) {
 //                testingData.cache();
 //            }
-            
+
+            // generating data for summary clusters
+            double sampleSize = (double) MLCoreServiceValueHolder.getInstance().getSummaryStatSettings()
+                    .getSampleSize();
+            double sampleFraction = sampleSize / (trainingData.count() - 1);
```
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java]: avoid division by 0.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java [hunk 2]
```diff
@@ -159,7 +168,38 @@ private ModelSummary buildKMeansModel(long modelID, JavaRDD<Vector> trainingData
...
+            JavaRDD<Vector> sampleData = null;
+            if (sampleFraction >= 1.0) {
+                sampleData = trainingData;
+            }
+            // Use ramdomly selected sample fraction of rows if number of records is > sample fraction
+            else {
+                sampleData = trainingData.sample(false, sampleFraction);
+            }
+
```
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:187]: At this point, please unpersist training data (as it is no more used) and cache sampleData.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java [hunk 3]
```diff
@@ -145,21 +154,58 @@ public MLModel build() throws MLModelBuilderException {
...
+            trainingData.unpersist(); // since no more used
+            sampleData.cache();
+
+            // Populate cluster points list with predicted clusters and features
+            List<Tuple2<Integer, Vector>> kMeansPredictions = kMeansModel.predict(sampleData).zip(sampleData).collect();
+            List<ClusterPoint> clusterPoints = new ArrayList<ClusterPoint>();
+
+            for (Tuple2<Integer, org.apache.spark.mllib.linalg.Vector> kMeansPrediction : kMeansPredictions) {
```
Comment (SupunS) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:195]: Does it need to specify the complete class path?

Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:195]: fixed with PR https://github.com/wso2/carbon-ml/pull/200

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/summary/ClusterModelSummary.java [hunk 4]
```diff
@@ -19,10 +19,12 @@
...
 
 import org.wso2.carbon.ml.commons.constants.MLConstants;
+import org.wso2.carbon.ml.commons.domain.ClusterPoint;
 import org.wso2.carbon.ml.commons.domain.ModelSummary;
 
 import java.io.Serializable;
 import java.util.Arrays;
+import java.util.List;
 
 public class ClusterModelSummary implements ModelSummary, Serializable {
```
Comment (lafernando) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/summary/ClusterModelSummary.java:29]: Add class level comments. Fix in all.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java [hunk 5]
```diff
@@ -90,6 +96,9 @@ public MLModel build() throws MLModelBuilderException {
             Workflow workflow = context.getFacts();
             long modelId = context.getModelId();
             ModelSummary summaryModel = null;
+            SortedMap<Integer, String> includedFeatures = MLUtils
```
Comment (prabathabey) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:99]: Code to interfaces, not concrete types, to be able to improve maintainability of the code.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java [hunk 6]
```diff
@@ -119,7 +128,7 @@ public MLModel build() throws MLModelBuilderException {
             UNSUPERVISED_ALGORITHM unsupervised_algorithm = UNSUPERVISED_ALGORITHM.valueOf(workflow.getAlgorithmName());
             switch (unsupervised_algorithm) {
             case K_MEANS:
-                summaryModel = buildKMeansModel(modelId, trainingData, testingData, workflow, mlModel);
+                summaryModel = buildKMeansModel(modelId, trainingData, testingData, workflow, mlModel,includedFeatures);
```
Comment (prabathabey) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:131]: Fix formatting issues.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java [hunk 7]
```diff
@@ -145,21 +154,58 @@ public MLModel build() throws MLModelBuilderException {
...
-            // remove from cache
-            trainingData.unpersist();
+
             // add test data to cache - test data is not used as of now
 //            if (testingData != null) {
 //                testingData.cache();
 //            }
-            
+
+            // generating data for summary clusters
```
Comment (prabathabey) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:170]: Use proper linguistics for comments. Use capitals to being the sentences.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java [hunk 8]
```diff
@@ -145,21 +154,58 @@ public MLModel build() throws MLModelBuilderException {
...
+            double sampleSize = (double) MLCoreServiceValueHolder.getInstance().getSummaryStatSettings()
+                    .getSampleSize();
+
+            double sampleFraction;
+            if(trainingData.count() != 1) { //avoiding division by 0
```
Comment (prabathabey) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:175]: Fix formatting issues.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java [hunk 9]
```diff
@@ -145,21 +154,58 @@ public MLModel build() throws MLModelBuilderException {
...
+                sampleFraction = sampleSize / (trainingData.count() - 1);
+            } else{
```
Comment (prabathabey) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:177]: More formatting issues.

File: apps/ml/site/analysis/analysis.jag [hunk 10]
```diff
@@ -320,15 +320,15 @@ include("../../includes/tenantAware.jag");
                                                 '</td>' +
                                                 '<td class="c-right">' +
                                                     '<div id="accordion-header-action-' + modelId + '">';
-                                                    // remove view and and predict links for k-means
-                                                    if(modelStatus == 'Complete' && sanitize(analysisModels[i].modelSummary.algorithm) != 'K_MEANS') {
+                                                    // remove predict links for k-means
+                                                    if(modelStatus == 'Complete') {
```
Comment (Kishanthan) [inline apps/ml/site/analysis/analysis.jag:324]: Formatting issues.

File: apps/ml/site/analysis/analysis.jag [hunk 11]
```diff
@@ -424,18 +424,15 @@ include("../../includes/tenantAware.jag");
...
-                                            $('<a href="#" id="view-model" class="cu-reg-btn btn-view" data-modelname="' + modelName + '" data-modelid="'+ modelId + '"><span>View</span></a>').insertBefore('#accordion-header-action-' + modelId + ' > #download-model');
-                                        }
-                                        // remove predict for collaborative filtering
-                                        if( (sanitize(models[i].modelSummary.algorithm) != 'COLLABORATIVE_FILTERING') && (sanitize(models[i].modelSummary.algorithm) != 'COLLABORATIVE_FILTERING_IMPLICIT') ){
-                                            $('<a href="#" id="predict" class="cu-reg-btn btn-predict" data-modelname="' + modelName + '" data-modelid="' + modelId + '"><span>Predict</span></a>').insertBefore('#accordion-header-action-' + modelId + ' > #download-model');
-                                        }
+                  
+                                    // remove view from k-means anomaly detection with unlabeled data
+                                    if(sanitize(models[i].modelSummary.algorithm) != 'K_MEANS_ANOMALY_DETECTION_WITH_UNLABELED_DATA'){
```
Comment (Kishanthan) [inline apps/ml/site/analysis/analysis.jag:429]: Formatting issues.

File: apps/ml/site/analysis/analysis.jag [hunk 12]
```diff
@@ -424,18 +424,15 @@ include("../../includes/tenantAware.jag");
...
+                                        $('<a href="#" id="view-model" class="cu-reg-btn btn-view" data-modelname="' + modelName + '" data-modelid="'+ modelId + '"><span>View</span></a>').insertBefore('#accordion-header-action-' + modelId + ' > #download-model');
                                     }
+                                    // remove predict for collaborative filtering and K-Means
+                                    if( (sanitize(models[i].modelSummary.algorithm) != 'COLLABORATIVE_FILTERING') && (sanitize(models[i].modelSummary.algorithm) != 'COLLABORATIVE_FILTERING_IMPLICIT') && (sanitize(models[i].modelSummary.algorithm) != 'K_MEANS')  ){
```
Comment (Kishanthan) [inline apps/ml/site/analysis/analysis.jag:433]: Formatting issues.

File: apps/ml/site/analysis/view-model.jag [hunk 13]
```diff
@@ -622,6 +622,15 @@ $( document ).ready(function() {
                  $("#model-summary-title").append(" [MSE: " + res.meanSquaredError.toExponential(2) + "]");
 
             }
+            else if(algorithm == 'K_MEANS'){
```
Comment (Kishanthan) [inline apps/ml/site/analysis/view-model.jag:625]: Formatting issues.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java (hunk 5 above)
Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:99]: fixed with https://github.com/wso2/carbon-ml/pull/201/

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java (hunk 6 above)
Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:131]: fixed with https://github.com/wso2/carbon-ml/pull/201/

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java (hunk 7 above)
Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:170]: fixed with https://github.com/wso2/carbon-ml/pull/201/

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java (hunk 8 above)
Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:175]: fixed with https://github.com/wso2/carbon-ml/pull/201/

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java (hunk 9 above)
Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/UnsupervisedSparkModelBuilder.java:177]: fixed with https://github.com/wso2/carbon-ml/pull/201/

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/summary/ClusterModelSummary.java (hunk 4 above)
Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/summary/ClusterModelSummary.java:29]: fixed with https://github.com/wso2/carbon-ml/pull/201/

File: apps/ml/site/analysis/view-model.jag (hunk 13 above)
Comment (fazlan-nazeem) [inline apps/ml/site/analysis/view-model.jag:625]: fixed with https://github.com/wso2/carbon-ml/pull/202

File: apps/ml/site/analysis/analysis.jag (hunk 12 above)
Comment (fazlan-nazeem) [inline apps/ml/site/analysis/analysis.jag:433]: fixed with https://github.com/wso2/carbon-ml/pull/202

File: apps/ml/site/analysis/analysis.jag (hunk 11 above)
Comment (fazlan-nazeem) [inline apps/ml/site/analysis/analysis.jag:429]: fixed with https://github.com/wso2/carbon-ml/pull/202

File: apps/ml/site/analysis/analysis.jag (hunk 10 above)
Comment (fazlan-nazeem) [inline apps/ml/site/analysis/analysis.jag:324]: fixed with https://github.com/wso2/carbon-ml/pull/202

**Pull request merged**